bolt-modules/boltlib/lib/puppet/functions/background.rb

```diff
@@ -31,7 +33,10 @@ def background(scope, name = nil, &block)
     executor = Puppet.lookup(:bolt_executor)
     executor.report_function_call(self.class.name)
 
-    executor.create_future(scope: scope, name: name) do |newscope|
+    # Use the Ruby object ID of the nested scopes to identify an individual
+    # plan invocation. This has to include
```

Comment:
```
    # Use the Ruby object ID of the nested scopes to identify an individual
    # plan invocation. Puppet scopes are a stack (not hierarchy), where there's a
    # top-level scope object with several 'layers' of scope underneath it. The
    # ephemeral instance variable contains the stack. The plan invocation ID
    # needs to include the object IDs for the entire scope stack because the
    # top-scope ID is persistent across new plan invocations, and additionally
    # new scopes are created during the course of a plan invocation. By
    # ensuring the scope stack for this plan invocation matches at least the
    # beginning of the scope stack in another plan function call, we can assert
    # that those functions belong in the same plan invocation.
```
I think I just didn't want to write it all out if this wasn't the solution we ended up using :P